### ISSUE TITLE
`insertCss` constraint

### ISSUE THREAD

**Issue by Rastopyr** [kriasoft/isomorphic-style-loader#60]
Aloha everyone,

during test coverage of my component i found the problem with shallow rendering.
For example code using [`enzyme`](http://airbnb.io/enzyme/):
```es6
const StyledComponent = withStyles('')(
  class Foo extends Component {
    render() {
      return <div />;
    }
  }
);

expect(shallow(<StyledComponent />).html()).to.equal('<div></div>');
```

After execute this code, test return error: `Cannot read property 'apply' of undefined`.
I found the cause of this error, and change `withStyles` module for fix error while maintaining backward compatibility.

About reason of error. I think this `insertCss` as required property in `propTypes` is unjustified constraint, because when you need create isolated of component you can't do this.

What you think about this issue?

Thanks!

**P.S.** if maintainer or community decide that actually problem, i can do PR for resolve this issue.

**Comment by dgutierrezHuge** [comment]
Hi, @Rastopyr.

I'm going through the same issue, could you tell me please how you fixed it?

**Comment by Rastopyr** [comment]
Of course, i fork loader and change required contextTypes to optional and add check for `this.context.insertCss` in `componentWillMount` hook.

You can see more details in fork:
[https://github.com/Nebo15/isomorphic-style-loader](https://github.com/Nebo15/isomorphic-style-loader)
[commit](https://github.com/Nebo15/isomorphic-style-loader/commit/68d8a79d96d1fdb9c1241730bc0573c482d7fa61)

**Comment by dgutierrezHuge** [comment]
@Rastopyr Thank you so much. I'll try the same solution.

**Comment by Rastopyr** [comment]
@dgutierrezHuge np 👍

**Comment by frenzzy** [comment]
Maybe `withStyles` higher-order component should not require `context.insertCss` and just warn if it was not provided, ref #63

@koistya what do you think?

**Comment by julykaz** [comment]
@koistya, is #63 going to be merged in near future or we need to use forked version with the fix?

**Comment by kmx411** [comment]
I'm unable to use react storybook because of this, would love to see it merged :)

**Comment by QuentinDejean** [comment]
Same here, thanks guys!

**Comment by samkelleher** [comment]
@kmx411 It's possible to use storybook whilst also using this loader, you have to add a global decorator in storybook config with a component that just stubs out the missing `insertCss` function in this example `ContextProvider.js`

```
import React, { Component, PropTypes } from 'react';

export default class ContextProvider extends Component {

    static propTypes = {
        children: PropTypes.node.isRequired,
        context: PropTypes.object.isRequired
    };

    static contextTypes = {
        insertCss: PropTypes.func
    };

    static childContextTypes = {
        insertCss: PropTypes.func.isRequired
    };

    getChildContext() {
        return this.props.context;
    }

    render() {
        return this.props.children;
    }

}
```

And use in `.storybook/config.js` as follows:

```
import React from 'react';
import { configure, addDecorator } from '@kadira/storybook';
import ContextProvider from './ContextProvider';

function loadStories() {...}

const context = {
    insertCss: () => {

    }
};

addDecorator(story => (
    <ContextProvider context={context}>
        {story()}
    </ContextProvider>
));

configure(loadStories, module);
```

Once you have this,  the undefined error will go away since `insertCss` now exists.

**Comment by pulkitnandan** [comment]
@samkelleher undefined error do sail away but react-starter-kit's used 'isomorphic-style-loader/lib/withStyles' doesn't work and every className is stripped. POC @ https://github.com/pulkitnandan/storyBookReactStarterKit

**Comment by dphrag** [comment]
If you have insertCss as a noop, then you wouldn't get your styles, so it kind of defeats the purpose of storybook.

If you do this instead it should magically work

  insertCss: (...styles) => {
    const removeCss = styles.map(x => x._insertCss());
    return () => {
      removeCss.forEach(f => f());
    };
  },

*(taken from: https://gist.github.com/alopes/982973c21519291f909c4e076a37f984)
